File: labs/mathlogic/lab3/lab3_1.py
from typing import Optional, List

class TreeNode:
    def __init__(self, data: int):
        self.data: int = data
        self.left: Optional['TreeNode'] = None
        self.right: Optional['TreeNode'] = None

class BinarySearchTree:
    def __init__(self):
        self.root: Optional[TreeNode] = None
        self._size: int = 0

    def _is_empty(self):
        return self._size == 0

    def _search_insert(self, key: int, node: TreeNode) -> TreeNode:
        if key < node.data:
            if node.left is None:
                return node
            return self._search_insert(key, node.left)
        if key > node.data:
            if node.right is None:
                return node
            return self._search_insert(key, node.right)
        return node

    def _direction_search(self, node: TreeNode, direction: str) -> TreeNode:
        if direction == "left":
            if node.left is None:
                return node
            return self._direction_search(node.left, direction)
        elif direction == "right":
            if node.right is None:
                return node
            return self._direction_search(node.right, direction)
        
        return node

    def _inorder(self, node: Optional[TreeNode], result: List[int]):
        if node is None:
            return
        self._inorder(node.left, result)
        result.append(node.data)
        self._inorder(node.right, result)

    def _preorder(self, node: Optional[TreeNode], result: List[int]):
        if node is None:
            return
        result.append(node.data)
        self._preorder(node.left, result)
        self._preorder(node.right, result)

    def _postorder(self, node: Optional[TreeNode], result: List[int]):
        if node is None:
            return
        self._postorder(node.left, result)
        self._postorder(node.right, result)
        result.append(node.data)

    def _delete_node(self, key: int, node: Optional[TreeNode]) -> Optional[TreeNode]:
        if node is None:
            return None
        
        if key < node.data:
            node.left = self._delete_node(key, node.left)
            return node
        
        if key > node.data:
            node.right = self._delete_node(key, node.right)
            return node
        
        if key == node.data:
            if node.left is None and node.right is None:
                self._size -= 1
                return None
            
            if node.left is None and node.right is not None:
                self._size -= 1
                return node.right
            
            if node.left is not None and node.right is None:
                self._size -= 1
                return node.left
            
            temp = self._direction_search(node.right, "left")
            node.data = temp.data
            node.right = self._delete_node(temp.data, node.right)
            return node
            
        return None

    def insert(self, key: int):
        if self._is_empty():
            self.root = TreeNode(key)
            self._size += 1
            return

        node = self._search_insert(key, self.root)
        
        if key > node.data:
            node.right = TreeNode(key)
            self._size += 1
            return
        
        if key < node.data:
            node.left = TreeNode(key)
            self._size += 1
            return

    def find(self, key: int) -> Optional[TreeNode]:
        if self._is_empty():
            return None
        
        node = self._search_insert(key, self.root)
        if node.data != key:
            return None
        return node

    def min(self):
        if self._is_empty():
            raise Exception("Tree is empty!")
        return self._direction_search(self.root, "left")

    def max(self):
        if self._is_empty():
            raise Exception("Tree is empty!")
        return self._direction_search(self.root, "right")

    def traverse_inorder(self):
        result: List[int] = []
        if self._is_empty():
            return result
        self._inorder(self.root, result)
        return result

    def traverse_preorder(self):
        result: List[int] = []
        if self._is_empty():
            return result
        self._preorder(self.root, result)
        return result

    def traverse_postorder(self):
        result: List[int] = []
        if self._is_empty():
            return result
        self._postorder(self.root, result)
        return result

    def delete(self, key: int):
        if self._is_empty():
            raise Exception("Tree is empty!")
        self.root = self._delete_node(key, self.root)

File: labs/mathlogic/lab3/lab3_1.ts
export class TreeNode {
  data: number;
  left: TreeNode | null = null;
  right: TreeNode | null = null;

  constructor(data: number) {
    this.data = data;
  }
}

export class BinarySearchTree {
  root: TreeNode | null = null;
  private _size: number = 0;

  private _is_empty() {
    return this._size == 0;
  }

  private _searchInsert(key: number, node: TreeNode) {
    if (key < node.data) {
      if (!node.left) {
        return node;
      }
      return this._searchInsert(key, node.left);
    }
    if (key > node.data) {
      if (!node.right) {
        return node;
      }
      return this._searchInsert(key, node.right);
    }
    return node;
  }

  private _directionSearch(node: TreeNode, direction: "left" | "right") {
    switch (direction) {
      case "left":
        if (!node.left) return node;
        return this._directionSearch(node.left, direction);
      case "right":
        if (!node.right) return node;
        return this._directionSearch(node.right, direction);
    }
  }

  private _inorder(node: TreeNode | null, result: number[]) {
    if (!node) return;
    this._inorder(node.left, result);
    result.push(node.data);
    this._inorder(node.right, result);
  }

  private _preorder(node: TreeNode | null, result: number[]) {
    if (!node) return;
    result.push(node.data);
    this._preorder(node.left, result);
    this._preorder(node.right, result);
  }

  private _postorder(node: TreeNode | null, result: number[]) {
    if (!node) return;
    this._postorder(node.left, result);
    this._postorder(node.right, result);
    result.push(node.data);
  }

  private _deleteNode(key: number, node: TreeNode | null): TreeNode | null {
    if (!node) return null;
    if (key < node.data) {
      node.left = this._deleteNode(key, node.left);
      return node;
    }
    if (key > node.data) {
      node.right = this._deleteNode(key, node.right);
      return node;
    }
    if (key == node.data) {
      if (!node.left && !node.right) {
        this._size--;
        return null;
      }
      if (!node.left && !!node.right) {
        this._size--;
        return node.right;
      }
      if (!!node.left && !node.right) {
        this._size--;
        return node.left;
      }
      const temp = this._directionSearch(node.right, "left");
      node.data = temp.data;
      node.right = this._deleteNode(temp.data, node.right);
      return node;
    }
  }

  insert(key: number) {
    if (this._is_empty()) {
      this.root = new TreeNode(key);
      this._size++;
      return;
    }
    const node = this._searchInsert(key, this.root);
    if (key > node.data) {
      node.right = new TreeNode(key);
      this._size++;
      return;
    }
    if (key < node.data) {
      node.left = new TreeNode(key);
      this._size++;
      return;
    }
  }

  find(key: number): TreeNode | null {
    if (this._is_empty()) {
      return null;
    }
    const node = this._searchInsert(key, this.root);
    if (node.data != key) return null;
    return node;
  }

  min() {
    if (this._is_empty()) {
      throw new Error("Tree is empty!");
    }
    return this._directionSearch(this.root, "left");
  }

  max() {
    if (this._is_empty()) {
      throw new Error("Tree is empty!");
    }
    return this._directionSearch(this.root, "right");
  }

  traverse_inorder() {
    const result: number[] = [];
    if (this._is_empty()) {
      return result;
    }
    this._inorder(this.root, result);
    return result;
  }

  traverse_preorder() {
    const result: number[] = [];
    if (this._is_empty()) {
      return result;
    }
    this._preorder(this.root, result);
    return result;
  }

  traverse_postorder() {
    const result: number[] = [];
    if (this._is_empty()) {
      return result;
    }
    this._postorder(this.root, result);
    return result;
  }

  delete(key: number) {
    if (this._is_empty()) {
      throw new Error("Tree is empty!");
    }
    this.root = this._deleteNode(key, this.root);
  }
}
